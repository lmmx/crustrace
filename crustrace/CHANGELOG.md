# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.4](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.3...crustrace-v0.1.4) - 2025-06-09

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.3](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.2...crustrace-v0.1.3) - 2025-06-08

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.2](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.1...crustrace-v0.1.2) - 2025-06-07

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.1](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.0...crustrace-v0.1.1) - 2025-06-07

### <!-- 2 -->Bug Fixes

- amend README metadata

### <!-- 9 -->Other

- license

## [0.1.0](https://github.com/lmmx/crustrace/releases/tag/crustrace-v0.1.0) - 2025-05-27

### <!-- 1 -->Features

- support impl Struct methods
- Initial commit

### <!-- 5 -->Refactor

- rename 'trace all' to 'omni', it's cleaner
- merge crustrace-attributes crate into crustrace-core ([#3](https://github.com/lmmx/crustrace/pull/3))

### <!-- 8 -->Styling

- lint

### <!-- 9 -->Other

- prepare for publishing crustrace crate
- set up git hooks
