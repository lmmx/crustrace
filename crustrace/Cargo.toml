[package]
description = "Function and module-level procedural macro attributes to instrument functions with tracing"
documentation = "https://docs.rs/crustrace"
edition = "2024"
homepage = "https://github.com/lmmx/crustrace"
license = "MIT OR Apache-2.0"
name = "crustrace"
readme = "../README.md"
repository = "https://github.com/lmmx/crustrace"
version = "0.1.0"

[lib]
proc-macro = true

[dependencies]
crustrace-core = { workspace = true }
proc-macro2.workspace = true

[dev-dependencies]
cargo-husky = { workspace = true, features = ["user-hooks"] }
insta.workspace = true
rust-format.workspace = true
tracing.workspace = true
tracing-subscriber.workspace = true
