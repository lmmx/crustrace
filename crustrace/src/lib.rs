use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;

#[proc_macro_attribute]
pub fn trace_all(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input2: TokenStream2 = input.into();
    let output = crustrace_core::trace_all_impl(input2);
    output.into()
}