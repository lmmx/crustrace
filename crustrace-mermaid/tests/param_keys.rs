#[cfg(test)]
mod param_keys_tests {
    use crustrace::instrument;
    use crustrace_mermaid::*;
    use tracing::subscriber::set_default;
    use tracing_subscriber::prelude::*;

    #[instrument]
    fn inner(x: i32, y: i32) -> i32 {
        x + y
    }

    #[instrument]
    fn outer(a: i32, b: i32) -> i32 {
        let r1 = inner(a + 1, b / 10);
        let r2 = inner(a * 2, b / 20);
        r1 + r2
    }

    #[test]
    fn snapshot_single_node() {
        let layer = MermaidLayer::new()
            .with_params_mode(ParamRenderMode::SingleNode)
            .without_auto_flush();
        let subscriber = tracing_subscriber::registry().with(layer.clone());
        let _guard = set_default(subscriber);

        outer(10, 20);

        insta::assert_snapshot!(layer.render());
    }

    #[test]
    fn snapshot_single_node_grouped() {
        let layer = MermaidLayer::new()
            .with_params_mode(ParamRenderMode::SingleNodeGrouped)
            .without_auto_flush();
        let subscriber = tracing_subscriber::registry().with(layer.clone());
        let _guard = set_default(subscriber);

        outer(10, 20);

        insta::assert_snapshot!(layer.render());
    }
}
