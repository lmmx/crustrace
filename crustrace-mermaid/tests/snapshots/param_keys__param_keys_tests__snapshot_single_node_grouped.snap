---
source: crustrace-mermaid/tests/param_keys.rs
expression: layer.render()
---
flowchart TD
F1["outer()"]:::func
PK1["outer(params):
• a = 10
• b = 20"]:::data
F1 -.-> PK1
F2["inner()"]:::func
PK2["inner(params):
• x = 11
• y = 2"]:::data
F2 -.-> PK2
  F1 --> F2
F3["inner()"]:::func
PK3["inner(params):
• x = 20
• y = 1"]:::data
F3 -.-> PK3
  F1 --> F3
subgraph ParamsKey["Parameters"]
  direction TB
  PK1
  PK2
  PK3
end

classDef func fill:#c6f6d5,stroke:#2f855a,stroke-width:2px,color:#22543d;
classDef data fill:#bee3f8,stroke:#2b6cb0,stroke-width:1.5px,color:#1a365d;
classDef params fill:none,stroke:#e53e3e,stroke-width:2px,color:#742a2a;
class PK1,PK2,PK3 params;
